express_relay/programs/express_relay/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+        let protocol_config = &ctx.accounts.protocol_config;
+        let protocol_config_account_info = protocol_config.to_account_info();
+        if protocol_config_account_info.data_len() > 0 {
+            let account_data = &mut &**protocol_config_account_info.try_borrow_data()?;
+            let protocol_config_data = ConfigProtocol::try_deserialize(account_data)?;
+            split_protocol = protocol_config_data.split;
+        } else {
+            split_protocol = split_protocol_default;
+        }
+
+        let fee_protocol = bid_amount * split_protocol / FEE_SPLIT_PRECISION;
+        if fee_protocol > bid_amount {
+            return err!(ErrorCode::FeesHigherThanBid);
+        }
+
+        let fee_relayer = bid_amount.saturating_sub(fee_protocol) * split_relayer / FEE_SPLIT_PRECISION;
+        if fee_relayer.checked_add(fee_protocol).ok_or(ProgramError::ArithmeticOverflow)? > bid_amount {
+            return err!(ErrorCode::FeesHigherThanBid);
+        }
+
+        let protocol = &ctx.accounts.protocol;
+        let fee_receiver_protocol = &ctx.accounts.fee_receiver_protocol;
+        if protocol.executable {
+            validate_pda(fee_receiver_protocol.key, protocol.key, &[SEED_EXPRESS_RELAY_FEES])?;
+        } else {
+            assert_eq!(protocol.key, fee_receiver_protocol.key);
+        }
+
+        let balance_fee_receiver_protocol = fee_receiver_protocol.lamports();
+        let rent_fee_receiver_protocol = Rent::get()?.minimum_balance(0);
+        if balance_fee_receiver_protocol+fee_protocol < rent_fee_receiver_protocol {
+            return err!(ErrorCode::InsufficientProtocolFeeReceiverRent);
+        }
+
+        let fee_receiver_relayer = &ctx.accounts.fee_receiver_relayer;
+        let balance_fee_receiver_relayer = fee_receiver_relayer.lamports();
+        let rent_fee_receiver_relayer = Rent::get()?.minimum_balance(0);
+        if balance_fee_receiver_relayer+fee_relayer < rent_fee_receiver_relayer {
+            return err!(ErrorCode::InsufficientRelayerFeeReceiverRent);
+        }
+
+        transfer_lamports_cpi(
+            &searcher.to_account_info(),
+            &fee_receiver_protocol.to_account_info(),
+            fee_protocol,
+            ctx.accounts.system_program.to_account_info()
+        )?;
+        transfer_lamports_cpi(
+            &searcher.to_account_info(),
+            &fee_receiver_relayer.to_account_info(),
+            fee_relayer,
+            ctx.accounts.system_program.to_account_info()
+        )?;
+        // send the remaining balance from the bid to the express relay metadata account
+        transfer_lamports_cpi(
+            &searcher.to_account_info(),
+            &express_relay_metadata.to_account_info(),
+            bid_amount.saturating_sub(fee_protocol).saturating_sub(fee_relayer),
+            ctx.accounts.system_program.to_account_info()
+        )?;
+
+        Ok(())
+    }
+
+    pub fn check_permission(ctx: Context<CheckPermission>) -> Result<()> {
```

Comment:
have you considered making this an sdk method an not a cpi? This doesn't need to be a CPI.

Reply:
we want integrated contracts to have a check for permissioning within their program logic, that way express relay permissioning cannot be foregone

Reply:
What I mean is that the integrated program can perform these checks without CPIng

Reply:
one advantage of the CPI is that the DAO controls the code, so we can upgrade it later. Harder to do that if it's in a client SDK used by a bunch of different programs.